samples/decider/src/shoppingCarts/shoppingCart.ts

```diff
@@ -0,0 +1,309 @@
+//////////////////////////////////////
+/// Shopping Carts
+//////////////////////////////////////
+
+import {
+  addProductItem,
+  assertProductItemExists,
+  ProductItem,
+  removeProductItem,
+} from './productItem';
+import { Decider } from '#core/decider';
+
+//////////////////////////////////////
+/// Events
+//////////////////////////////////////
+
+export type ShoppingCartEvent =
+  | {
+      type: 'ShoppingCartOpened';
+      data: {
+        shoppingCartId: string;
+        clientId: string;
+        openedAt: string;
+      };
+    }
+  | {
+      type: 'ProductItemAddedToShoppingCart';
+      data: {
+        shoppingCartId: string;
+        productItem: ProductItem;
+      };
+    }
+  | {
+      type: 'ProductItemRemovedFromShoppingCart';
+      data: {
+        shoppingCartId: string;
+        productItem: ProductItem;
+      };
+    }
+  | {
+      type: 'ShoppingCartConfirmed';
+      data: {
+        shoppingCartId: string;
+        confirmedAt: string;
+      };
+    }
+  | {
+      type: 'ShoppingCartCanceled';
+      data: {
+        shoppingCartId: string;
+        canceledAt: string;
+      };
+    };
+
+//////////////////////////////////////
+/// Entity/State
+//////////////////////////////////////
+
+export type ShoppingCart =
+  | {
+      status: 'Empty';
+    }
+  | {
+      status: 'Pending';
+      id: string;
+      clientId: string;
+      productItems: ProductItem[];
+    }
+  | {
+      status: 'Confirmed';
+      id: string;
+      clientId: string;
+      productItems: ProductItem[];
+      confirmedAt: Date;
+    }
+  | {
+      status: 'Canceled';
+      id: string;
+      clientId: string;
+      productItems: ProductItem[];
+      canceledAt: Date;
+    };
+
```

Comment:
Hi there! Came from the article, and I had some suggestions that maybe you would find insightful

Utilize [TypeScript's discriminator types](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#discriminating-unions) to get true type-safety on your `ShoppingCart` type via type-narrowing, which can guarantee these conditional properties are there (assuming the interface is respected, which you could really guarantee via JSON validation) after you check the status
```suggestion
export enum ShoppingCartType {
  Empty = 'Empty',
  Pending = 'Pending',
  Confirmed = 'Confirmed',
  Canceled = 'Canceled',
}

type ShoppingCartEmpty = {
  status: ShoppingCartType.Empty;
};

type ShoppingCartBase = {
  id: string;
  clientId: string;
  productItems: ProductItem[];
};

type ShoppingCartPending = ShoppingCartBase & {
  status: ShoppingCartType.Pending;
};
type ShoppingCartConfirmed = ShoppingCartBase & {
  status: ShoppingCartType.Confirmed;
  confirmedAt: Date;
};
type ShoppingCartCanceled = ShoppingCartBase & {
  status: ShoppingCartType.Canceled;
  canceledAt: Date;
};

export type ShoppingCart =
  | ShoppingCartEmpty
  | ShoppingCartPending
  | ShoppingCartConfirmed
  | ShoppingCartCanceled;
```

Seems like you definitely understand the union operator (`|`) for TypeScript types -- but the intersection operator (`&`) is just as powerful as well (especially when they are used together like in this example).

Other changes in this code is making status an enum. The enum is exactly the same thing as what you're doing right now at runtime, but the enum helps by making all references to these constant string values (i.e. `"Empty"`) explicit, which is good for traceability.

This suggestion also applies to your other types as well, so this discriminator type pattern could be applied to all of your types

Reply:
Thank you, that's a fair comment, I'm aware of that, and that's how I'm using it in the other samples. Here, I wanted to keep things as simple as possible, thus the single type.

I'm also not so fond of the base types like `ShoppingCartBase`. I prefer to have explicit that they're different states. A bit of copy-paste won't harm anyone but will make it harder to try to uniform what's different accidentally. I'm treating those types as form of documentation. I prefer to keep those states clearly stating what data they have without the need to check also other types.